src/org/tukaani/xz/XZOutputStream.java

```diff
@@ -319,6 +319,15 @@ public void updateFilters(FilterOptions[] filterOptions)
             throw new UnsupportedOptionsException("Changing filter options "
                     + "in the middle of a XZ Block not implemented");
 
+        setFiltersForNextBlock(filterOptions);
+    }
+
+    // This is called from updateFilters and from a constructor.
+    // Since 1.10, the constructor doesn't call the public function
+    // updateFilters because it might lead to this-escape if this class
+    // is extended and updateFilters is overridden.
+    private void setFiltersForNextBlock(FilterOptions[] filterOptions)
```

Comment:
`setFiltersForNextBlock` sets the member variables `filtersSupportFlushing` and `filters`. So a static function would need to return both. My first thought is that it could make the code more complex.

Reply:
It does - but only as a side effect of asking each of the FilterEncoder instances. So you could break this into 2 steps:

`private static FilterEncoder[] buildEncoders(FilterOptions[] options)`

and a second

```
private static boolean canAllFlush(FilterEncoder[] filters) {
  for (FilterEncoder filter : filters) {
    if (!filter.supportsFlushing()) {
      return false;
    }
  }
  return true;
}
```

It is slightly more code than current, but only marginally so and I suspect no measurable performance impact given how small the number of filters can be.

Reply:
It's not hot code so performance cannot be affected.

I understand that there would be two static functions. Then the constructor and `updateFilters` would both need the same two lines:

    filters = buildEncoders(filterOptions);
    filtersSupportFlushing = canAllFlush(filters);

This sounds mildly error-prone. If, for some reason, the same functionality was needed in a third location, one would need to copypaste both lines. Copying only the first one would lead to subtle bugs because it would often appear to work fine.

It feels safer and clearer to keep such dependent pieces behind a single function call. That is, having those two static functions could be OK if a non-static `setFiltersForNextBlock` or such would still exist with the above two lines. But the current function is quite short and not too ugly, so I still feel it's pretty good as it is.

Reply:
Yes - I would think still having a member method `setFiltersForNextBlock` and then the constructor and that `setFiltersForNextBlock` would be the only 2 places calling the static methods.

It is kind of tricky dealing with member methods to mutate member state being called from a constructor on non-final classes.

Reply:
I understand the trickyness nowadays. When I wrote these, `javac` from OpenJDK 5 or 6 (or even 8) didn't warn about this-escape and I didn't understand the problem. Nowadays I wonder if it was good that such code was allowed to compile at all.

Calling private member functions is safe at least from this-escape point of view. I don't know if it is a bad pattern in some other ways still though.

My point was that if there was

    private void setFiltersForNextBlock(FilterOptions[] filterOptions)
            throws XZIOException {
        filters = buildEncoders(filterOptions);
        filtersSupportFlushing = canAllFlush(filters);
    }

then I don't see why the constructor shouldn't call it to avoid duplicating those two lines. What if later a third line is needed here and I forget to update the constructor. It's safest to do the complete thing in exactly one place. And if a constructor calls this function, I feel it's not worth it to add two static helper functions *in this case*.

While fixing the warning here is a small API/ABI change, I suppose it's fine enough. It takes pretty weird subclass to cause actual breakage. Here my point is that it would be more annoying to create `org.tukaani.xz2` because of these fixes. In that case I perhaps rather leave the this-escape warnings unfixed.

The filter options cases are more annoying as `LZMA2Options` has so many affected functions.